<!doctype html><html lang=en-us><head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width,initial-scale=1"><meta name=author content="Lukas Knuth"><meta name=description content="I recently answered a question on StackOverflow, asking if catching an Error would be reasonable in a particular case. The original question and my answer can be found here, however I felt that my answer could be extended with a more general discussion of the implications and additional examples."><meta property="og:description" value="I recently answered a question on StackOverflow, asking if catching an Error would be reasonable in a particular case. The original question and my answer can be found here, however I felt that my answer could be extended with a more general discussion of the implications and additional examples."><meta property="og:title" value="Catching Practice"><meta property="og:url" value=https://lknuth.dev/writings/catching_practice/><meta property="og:type" value=article><meta property="og:article:author" value="Lukas Knuth"><meta property="og:article:published_time" value=2012-07-06T20:40:00+0100><meta name=robots content="all,follow"><meta name=googlebot content="index,follow,snippet,archive"><title>Catching Practice | Lukas Knuth</title>
<link rel="stylesheet" href="/css/main.min.8e36eaaa6e050f395053719e62d46372876b276b94ce8375fab4aea4e8528a1e.css" integrity="sha256-jjbqqm4FDzlQU3GeYtRjcodrJ2uUzoN1+rSupOhSih4=" crossorigin="anonymous">
<link rel=canonical href=https://lknuth.dev/writings/catching_practice/><link rel=icon type=image/png href=https://lknuth.dev/images/favicon-96x96.png sizes=96x96><link rel=icon type=image/svg+xml href=https://lknuth.dev/images/favicon.svg><link rel="shortcut icon" href=https://lknuth.dev/images/favicon.ico><link rel=alternate type=application/rss+xml href=https://lknuth.dev/writings/index.xml title=Writings></head><body id=very-top class="flex flex-col h-screen max-w-6xl container mx-auto bg-pane break-words text-base text-writing md:text-lg antialiased hyphens-auto leading-relaxed font-sans"><header><div class="h-spaced mt-4 md:mt-6 mb-3"><a href=https://lknuth.dev/><div class="w-16 h-16 mr-3 md:mr-5 float-left bg-no-repeat bg-center bg-contain rounded-md" style=background-image:url(https://lknuth.dev/images/favicon.svg)></div><div class="hidden md:!block text-3xl font-bold"><span>Lukas Knuth</span><span class="pl-1 pr-3 tracking-[-0.3em]">//</span><span>Software Engineer</span></div><div class="md:hidden text-xl font-bold truncate text-clip"><span>Lukas Knuth</span><span class="pl-1 pr-2 tracking-[-0.3em]">//</span><span>SWE</span></div></a><nav><ol class="flex text-primary pt-1"><li class="md:text-base inline-block leading-4 mt-1 pr-2 mr-2 md:pr-3 md:mr-3 align-middle border-dotted border-r-2 last:border-none"><a href=/writings>Writings</a></li><li class="md:text-base inline-block leading-4 mt-1 pr-2 mr-2 md:pr-3 md:mr-3 align-middle border-dotted border-r-2 last:border-none"><a href=/about>About</a></li><li class="md:text-base inline-block leading-4 mt-1 pr-2 mr-2 md:pr-3 md:mr-3 align-middle border-dotted border-r-2 last:border-none"><a href=/cv>CV</a></li><li class="md:text-base inline-block leading-4 mt-1 pr-2 mr-2 md:pr-3 md:mr-3 align-middle border-dotted border-r-2 last:border-none"><a href=/now>Now</a></li></ol></nav></div></header><main class=grow><article class="h-spaced mt-4 md:mt-6" itemscope itemtype=http://schema.org/BlogPosting><h1 class="text-3xl font-light text-subtle" itemprop="name headline">Catching Practice</h1><div class="my-2 text-sm text-subtle"><time datetime="2012-07-06 20:40:00 +0100 +0100" itemprop=datePublished>2012-07-06</time></div><div class=prose itemprop=articleBody><p>I recently answered a question on StackOverflow, asking if catching an <code>Error</code> would be reasonable in a particular case. The original question and my answer can be found <a href=http://stackoverflow.com/q/11017304/717341 rel=external>here</a>, however I felt that my answer could be extended with a more general discussion of the implications and additional examples.</p><h2 id=clarification-throwable-error-exception><a href=#clarification-throwable-error-exception>Clarification: Throwable, Error, Exception</a></h2><p>In Java, errors and exceptions (which are the main types) are thrown using the <code>throw</code>-keyword. Every class which extends the basic <code>java.lang.Throwable</code> can be thrown.</p><p>In the Standard Library, there are two classes which inherit from the basic <code>Throwable</code>-class: <code>Exception</code> and <code>Error</code>. Their semantics are explained in the respective documentation:</p><p><strong><a href=http://docs.oracle.com/javase/6/docs/api/java/lang/Error.html rel=external>Error</a></strong></p><blockquote><p>An <code>Error</code> is a subclass of <code>Throwable</code> that <strong>indicates serious problems that a reasonable application <em>should not</em> try to catch.</strong>
Most such errors are abnormal conditions. [&mldr;]</p></blockquote><p><strong><a href=http://docs.oracle.com/javase/6/docs/api/java/lang/Exception.html rel=external>Exception</a></strong></p><blockquote><p>The class <code>Exception</code> and its subclasses are a form of <code>Throwable</code> that <strong>indicates conditions that a reasonable application
<em>might</em> want to catch.</strong></p></blockquote><p>So, it appears, that it&rsquo;s generally a bad idea to catch an <code>Error</code> on purpose.</p><h2 id=why-do-we-distinguish><a href=#why-do-we-distinguish>Why do we distinguish?</a></h2><p>Additionally, exceptions are further divided into &ldquo;checked exceptions&rdquo; (which inherit from the basic <code>java.lang.Exception</code>-class) and &ldquo;runtime exceptions&rdquo; (which inherit from <code>java.lang.RuntimeException</code>).</p><p>The difference being that when a &ldquo;checked exception&rdquo; is thrown, it must be handled (by a &ldquo;try/catch&rdquo;-block), whereas a &ldquo;runtime exception&rdquo; <em>should not</em> be handled at all.</p><p>The reason why there are multiple different types of <code>Throwable</code> is, because they all have different origins:</p><ul><li>An <code>Error</code> represents a critical low-level problem (probably with the VM) which an application is <em>most likely</em> not able to recover from.</li><li>A &ldquo;checked exception&rdquo; indicates a high-level problem, which the application is <em>likely</em> to recover from.</li><li>A &ldquo;runtime exception&rdquo; indicates a programming error, which the application <em>should not</em> recover from.</li></ul><p>The keyword here really is <em>recover</em>.</p><h2 id=can-we-recover><a href=#can-we-recover>Can we recover?</a></h2><p>The question to ask when catching any kind of <code>Throwable</code> is always the following: <strong>Can my application recover from it?</strong> If the application <em>can</em> recover, it should catch. If it <em>can&rsquo;t</em> (or is most likely unable to) recover, it shouldn&rsquo;t.</p><p>From the (excellent) &ldquo;<a href=http://www.amazon.com/Effective-Java-Edition-Joshua-Bloch/dp/0321356683 rel=external>Effective Java - Second Edition</a>&rdquo; by Joshua Bloch, Chapter 9, Item 58:</p><blockquote><p>If a program throws an unchecked exception or an error, it is generally the case that recovery is impossible and
continued execution would do more harm than good.</p></blockquote><p>Now, the task is to determine whether we can recover or not. The type of <code>Throwable</code> gives us a hint, but it&rsquo;s really depending on the actual situation.</p><h2 id=examples><a href=#examples>Examples</a></h2><h3 id=error><a href=#error>Error</a></h3><p>This example is from the original question linked above:</p><blockquote><p>[&mldr;] I&rsquo;m currently loading a DLL that is not guaranteed to be on the PATH, and would like to switch to a
user-configured location in the case that it isn&rsquo;t.</p></blockquote><p>The code is given as follows:</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#8839ef>try</span> {
</span></span><span style=display:flex><span>    System.<span style=color:#1e66f5>loadLibrary</span>(<span style=color:#40a02b>&#34;HelloWorld&#34;</span>);
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span>(UnsatisfiedLinkError ule){
</span></span><span style=display:flex><span>    System.<span style=color:#1e66f5>load</span>(<span style=color:#40a02b>&#34;C:/libraries/HelloWorld.dll&#34;</span>);
</span></span><span style=display:flex><span>}
</span></span></code></pre></div><p>Quoting from the <code>java.lang.System</code>-<a href=http://docs.oracle.com/javase/6/docs/api/java/lang/System.html#loadLibrary%28java.lang.String%29 rel=external>JavaDoc</a>, the <code>UnsatisfiedLinkError</code> is thrown&mldr;</p><blockquote><p>Throws: UnsatisfiedLinkError - if the library does not exist.</p></blockquote><p>So, <strong>can we recover?</strong> The application tries to load the library from the <code>PATH</code>. If this does not succeed, there is a fallback location to look for the library. Hence, the application is able to recover from the error-condition.</p><p>So, in this example, it&rsquo;s perfectly reasonable to catch the <code>Error</code>. Whether or not the library is guaranteed to be at the fallback location is outside the scope of this example!</p><h3 id=runtimeexception><a href=#runtimeexception>RuntimeException</a></h3><p>Imagine you&rsquo;re trying to parse a CSV-file containing integers. Reading the file line by line using a <code>BufferedReader</code>, we get a string per line. If you want integers, you need to parse.</p><p>Unfortunately, it&rsquo;s not guaranteed that what you expect to be an integer in the CSV file is actually even a number. One possibility is, that the value is empty, in which case you want to use a default value instead:</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#8839ef>try</span> {
</span></span><span style=display:flex><span>    last_int <span style=color:#04a5e5;font-weight:700>=</span> Integer.<span style=color:#1e66f5>parseInt</span>(foo); <span style=color:#9ca0b0;font-style:italic>// Where foo is a string-value from the CSV</span>
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (NumberFormatException e){
</span></span><span style=display:flex><span>    last_int <span style=color:#04a5e5;font-weight:700>=</span> 1; <span style=color:#9ca0b0;font-style:italic>// As our default value</span>
</span></span><span style=display:flex><span>}
</span></span></code></pre></div><p>From the <a href=http://docs.oracle.com/javase/6/docs/api/java/lang/Integer.html#parseInt%28java.lang.String%29 rel=external>JavaDoc</a> of <code>java.lang.Integer#parseInt(String)</code>:</p><blockquote><p>Throws: NumberFormatException - if the string does not contain a parsable integer.</p></blockquote><p>And, believe it or not, an empty string is not a parsable integer:</p><pre><code>Exception in thread &quot;main&quot; java.lang.NumberFormatException: For input string: &quot;&quot;
    at java.lang.NumberFormatException.forInputString(Unknown Source)
    at java.lang.Integer.parseInt(Unknown Source)
</code></pre><p>So again, we&rsquo;re catching an exception which we are (by convention) not supposed to. In this case again, we&rsquo;re able to recover from it and can continue the execution without problems.</p><p>One could now argue that it might be a design-error to make the <code>NumberFormatException</code>-class a runtime exception, but as mentioned above, runtime exceptions indicate <em>programming errors</em>. And since it <em>is</em> possible to know if a string is a real integer <em>before</em> parsing it, it can be considered a programming error to not check first. For further reading, check <a href=http://stackoverflow.com/a/7205689/717341 rel=external>this SO question</a>.</p><p>Additionally, using checked exceptions extensively tends to clutter calling code with <code>try/catch</code> blocks or add ridiculous amounts of <code>throws</code> declarations on methods. Check the <a href=#design-advices>&ldquo;design advices&rdquo;</a> below for further reading.</p><h3 id=exception><a href=#exception>Exception</a></h3><p>If an application is likely to <em>not</em> recover from a thrown, <u>checked</u>, <code>Exception</code>, then making it a checked exception was a design-error in the first place.</p><p>None the less, such cases need to be handled, since the compiler wont let you compile the program without catching this exception.</p><p>Since we&rsquo;ve already established that we aren&rsquo;t able to recover from the <code>Exception</code> and should really stop the program, let&rsquo;s do so using a <code>RuntimeException</code> but give all available information by specifying the original exception as the cause:</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#8839ef>try</span> {
</span></span><span style=display:flex><span>    ifThisDoesntWorkWereScrewed();
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (UnrecoverableCheckedException e) {
</span></span><span style=display:flex><span>    <span style=color:#8839ef>throw</span> RuntimeException(<span style=color:#40a02b>&#34;Can&#39;t recover from this!&#34;</span>, e);
</span></span><span style=display:flex><span>}
</span></span></code></pre></div><p>This way the previously checked exception now behaves as it should, while no information is lost because it&rsquo;s stack-trace is included in the newly created <code>RuntimeException</code>. <strong>Note</strong> that this is a workaround and should ideally never be necessary!</p><h2 id=design-advices><a href=#design-advices>Design advices</a></h2><p>To round everything up, some advices if you&rsquo;re writing an API or library yourself.</p><h3 id=making-the-ball><a href=#making-the-ball>Making the ball</a></h3><p>As a general advice when creating a library/API: <strong>Favor the use of standard exceptions</strong>. Again, quoting from &ldquo;Effective Java - Second Edition&rdquo;, Chapter 9, Item 60:</p><blockquote><p>Reusing preexisting exceptions has several benefits. Chief among these, <strong>it makes your API easier to learn and use
because it matches established conventions</strong> with which programmers are already familiar. A close second is
that programs using your API are <strong>easier to read because they aren&rsquo;t cluttered with unfamiliar exceptions</strong>.
[&mldr;]</p></blockquote><p>But, aside from this, there are cases when you need to create your own exceptions because you want to express something more specific, or there is just no standard exception which meets your needs.</p><p>In these cases, do so, but <strong>never extend <code>Throwable</code> directly!</strong> It&rsquo;s not that the compiler wont let you (it will work like a &ldquo;checked exception&rdquo;, just like the <code>Exception</code>-class), the Java Language Specs do not address such throwables directly. So, you&rsquo;re going into <em>undefined behavior</em>, which should always be avoided.</p><p>Instead, extend specific subclasses of <code>Exception</code> or <code>RuntimeException</code> (to get a new &ldquo;checked-&rdquo; or &ldquo;unchecked-exception&rdquo; respectively). It&rsquo;s by convention not wrong to make your new <code>CSVParsingException</code> extend the basic <code>Exception</code>-class, but extending the more specific <code>ParseException</code> (which is a subclass of <code>Exception</code> - therefor a checked exception) makes for a cleaner picture.</p><p>Of course if you can&rsquo;t find any &ldquo;more specific&rdquo; subclasses, you&rsquo;ll need to extend the basic <code>Exception</code>- or <code>RuntimeException</code>-classes.</p><p>Also, be sure to <strong>add a constructor which takes a <code>Throwable cause</code>-parameter</strong> to preserve information from the originally thrown Exception when re-throwing (see further down). The basic <code>Exception</code>-class offers <a href=http://docs.oracle.com/javase/6/docs/api/java/lang/Exception.html#Exception%28java.lang.String,%20java.lang.Throwable%29 rel=external>such a constructor</a> which you can simply overwrite.</p><p>A word on extending <code>Error</code>: <u>Don&rsquo;t do it.</u> As mentioned above, errors are (by convention) reserved for low-level problems (with the VM). Instead, use a <code>RuntimeException</code>.</p><h3 id=throwing-the-ball><a href=#throwing-the-ball>Throwing the ball</a></h3><p><em>Think twice before you throw.</em> Checked exceptions are a powerful language-feature, because other than return types, you can force the developer to handle them. This can be nice but can also make the resulting code very bloated, as every thrown exception needs to be handled in it&rsquo;s own <code>catch</code>-block (yes, there is a Java 7 feature for that&mldr;).</p><p>So, thinking about whether to throw a checked or unchecked exception can make your API/library more pleasant to use. As a rule of thumb consider the following:</p><blockquote><p>The burden [of handling a checked exception] is justified <strong>if the exceptional condition cannot be prevented by
proper use of the API <em>and</em> the programmer using the API can take some useful action once confronted with the
exception</strong>. Unless both of these conditions hold, an unchecked exception is more appropriate.</p><p><em>From &ldquo;Effective Java - Second Edition&rdquo;, Chapter 9, Item 59</em></p></blockquote><p>Just like in the above &ldquo;Making the ball&rdquo; section, a short word on throwing <code>Error</code>: <u>Don&rsquo;t do it.</u> Since we already clarified that errors are reserved for low-level problems, there should be no need to throw them yourself.</p><h3 id=catching-the-ball><a href=#catching-the-ball>Catching the ball</a></h3><p>Exceptions tend to be ignored. For unchecked exceptions and errors, this might be desired (in most cases, see above), but it completely defeats the purpose of throwing a checked exception. I have seen this in various places (even in the <a href=http://docs.oracle.com/javase/6/docs/api/javax/swing/SwingWorker.html rel=external>Java Docs</a>):</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#9ca0b0;font-style:italic>// Bad practice example. DON&#39;T DO THIS!</span>
</span></span><span style=display:flex><span><span style=color:#8839ef>try</span> {
</span></span><span style=display:flex><span>    somethingThatThrows();
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (SomeException ignored){
</span></span><span style=display:flex><span>    <span style=color:#9ca0b0;font-style:italic>// Can&#39;t happen...</span>
</span></span><span style=display:flex><span>}
</span></span></code></pre></div><p>&mldr; until the day it does. Also bad:</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#9ca0b0;font-style:italic>// Bad practice example. DON&#39;T DO THIS!</span>
</span></span><span style=display:flex><span><span style=color:#8839ef>try</span> {
</span></span><span style=display:flex><span>    somethingThatTrows();
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (SpecificException e){
</span></span><span style=display:flex><span>    <span style=color:#8839ef>throw</span> <span style=color:#8839ef>new</span> OtherException(<span style=color:#40a02b>&#34;Doesn&#39;t work&#34;</span>);
</span></span><span style=display:flex><span>}
</span></span></code></pre></div><p>While this example <em>will</em> throw an exception, it <strong>totally discards any (useful) information from the previously thrown <code>SpecificException</code></strong>. It&rsquo;s not generally wrong to re-throw an exception, but you should either print the stack-trace of the original one, or (even better) include it as the cause in the new Exception. Every exception from the standard API has a constructor which accepts a <code>Throwable cause</code> as an argument (and your custom ones should, too). Use it!</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#8839ef>throw</span> <span style=color:#8839ef>new</span> OtherException(<span style=color:#40a02b>&#34;Doesn&#39;t work&#34;</span>, e);
</span></span></code></pre></div><p>And now, the worst possible approach:</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#9ca0b0;font-style:italic>// Bad practice example. DON&#39;T DO THIS!</span>
</span></span><span style=display:flex><span><span style=color:#8839ef>try</span> {
</span></span><span style=display:flex><span>    somethingThatThrows();
</span></span><span style=display:flex><span>    somethingElseThatThrows();
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (Exception ignore) {
</span></span><span style=display:flex><span>    <span style=color:#9ca0b0;font-style:italic>// Ignored.</span>
</span></span><span style=display:flex><span>}
</span></span></code></pre></div><p>The last example (sometimes referred to as &ldquo;<a href=http://stackoverflow.com/a/2308988/717341 rel=external>Pokemon Exception Handling</a>&rdquo;) is generally <strong>discarding <em>all</em> possibly thrown exceptions</strong> (including the unchecked ones). If something goes wrong in this piece of code, your program wont crash but it wont behave as expected either. And you won&rsquo;t get any information about it.</p><p>Exceptions are not your enemies. They are a helping hand from the runtime. So don&rsquo;t abandon them. If there is really nothing you can (or want) to do about an exception, <strong>at least</strong> print the stack-trace (when in developing stage) or log them anywhere (when in productive state). Otherwise, some fine day, you might suffer from the revenge of the forgotten ones.</p><h4 id=catching-generally><a href=#catching-generally>Catching generally</a></h4><p>Last but not least, an advice on catching more general errors: <u>Do it with extreme caution</u></p><p>Third party code in your application (such as libraries or even the standard library) might change with a new version of the library. A general catching block could swallow a newly added exception and suddenly your applications code (which wasn&rsquo;t even touched) stops working as expected.</p><p>The only thing you should do in generalized <code>catch</code>-blocks is logging. For example, the following is perfectly reasonable:</p><div class=highlight><pre tabindex=0 style=color:#4c4f69;background-color:#eff1f5;-moz-tab-size:2;-o-tab-size:2;tab-size:2><code class=language-java data-lang=java><span style=display:flex><span><span style=color:#8839ef>try</span> {
</span></span><span style=display:flex><span>    somethingThatThrows();
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (SpecificException e){
</span></span><span style=display:flex><span>    <span style=color:#9ca0b0;font-style:italic>// Handle the specific exception here</span>
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (OtherSpecificException e){
</span></span><span style=display:flex><span>    <span style=color:#9ca0b0;font-style:italic>// Handle the other specific exception here</span>
</span></span><span style=display:flex><span>} <span style=color:#8839ef>catch</span> (GeneralException e){
</span></span><span style=display:flex><span>    e.<span style=color:#1e66f5>printStackTrace</span>();
</span></span><span style=display:flex><span>    logger.<span style=color:#1e66f5>error</span>(e);
</span></span><span style=display:flex><span>}
</span></span></code></pre></div><p>The important part is, that the more specific exceptions need to be caught before the more general one. And when doing so, <code>GeneralException</code> should be as specific as you can make it.</p><h4 id=returning-from-finally><a href=#returning-from-finally>Returning from finally</a></h4><p>As you are handling exceptions in your code, you should remember to <strong>never return in a <code>finally</code>-block!</strong>. This can become a serious problem, as outlined in <a href=http://weblogs.java.net/blog/staufferjames/archive/2007/06/_dont_return_in.html rel=external>James Stauffer&rsquo;s article</a>:</p><blockquote><p>If you return in a finally block then <strong>any <code>Throwables</code> that aren&rsquo;t caught (in a catch block that is part of the
same try statement as the <code>finally</code> block) will be completely lost</strong>. The really bad part about this is that it
looks so innocent.<br>[&mldr;]<br>Note <strong>this also applies to other things that transfer control</strong> (continue, throw, etc).</p></blockquote><p>So keep an eye out for that.</p><h2 id=conclusion><a href=#conclusion>Conclusion</a></h2><ul><li>If you can recover from an exception/error, don&rsquo;t hesitate to catch it.</li><li>Prefer the default exceptions over creating your own.</li><li>When creating your own exceptions, don&rsquo;t extend <code>Throwable</code> <em>directly</em> and don&rsquo;t extend <code>Error</code>.</li><li>Many checked exceptions make the calling code bloated.</li><li>Never ignore or discard exceptions without <strong>at least</strong> logging them.</li><li>Never return from within a <code>finally</code>-block.</li></ul></div></article><div class="mt-8 mb-3 py-4 px-5 md:px-9" style="background-image:url(&#34;data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 304 304' width='304' height='304'%3E%3Cpath fill='%23c8c8c8' fill-opacity='0.5' d='M44.1 224a5 5 0 1 1 0 2H0v-2h44.1zm160 48a5 5 0 1 1 0 2H82v-2h122.1zm57.8-46a5 5 0 1 1 0-2H304v2h-42.1zm0 16a5 5 0 1 1 0-2H304v2h-42.1zm6.2-114a5 5 0 1 1 0 2h-86.2a5 5 0 1 1 0-2h86.2zm-256-48a5 5 0 1 1 0 2H0v-2h12.1zm185.8 34a5 5 0 1 1 0-2h86.2a5 5 0 1 1 0 2h-86.2zM258 12.1a5 5 0 1 1-2 0V0h2v12.1zm-64 208a5 5 0 1 1-2 0v-54.2a5 5 0 1 1 2 0v54.2zm48-198.2V80h62v2h-64V21.9a5 5 0 1 1 2 0zm16 16V64h46v2h-48V37.9a5 5 0 1 1 2 0zm-128 96V208h16v12.1a5 5 0 1 1-2 0V210h-16v-76.1a5 5 0 1 1 2 0zm-5.9-21.9a5 5 0 1 1 0 2H114v48H85.9a5 5 0 1 1 0-2H112v-48h12.1zm-6.2 130a5 5 0 1 1 0-2H176v-74.1a5 5 0 1 1 2 0V242h-60.1zm-16-64a5 5 0 1 1 0-2H114v48h10.1a5 5 0 1 1 0 2H112v-48h-10.1zM66 284.1a5 5 0 1 1-2 0V274H50v30h-2v-32h18v12.1zM236.1 176a5 5 0 1 1 0 2H226v94h48v32h-2v-30h-48v-98h12.1zm25.8-30a5 5 0 1 1 0-2H274v44.1a5 5 0 1 1-2 0V146h-10.1zm-64 96a5 5 0 1 1 0-2H208v-80h16v-14h-42.1a5 5 0 1 1 0-2H226v18h-16v80h-12.1zm86.2-210a5 5 0 1 1 0 2H272V0h2v32h10.1zM98 101.9V146H53.9a5 5 0 1 1 0-2H96v-42.1a5 5 0 1 1 2 0zM53.9 34a5 5 0 1 1 0-2H80V0h2v34H53.9zm60.1 3.9V66H82v64H69.9a5 5 0 1 1 0-2H80V64h32V37.9a5 5 0 1 1 2 0zM101.9 82a5 5 0 1 1 0-2H128V37.9a5 5 0 1 1 2 0V82h-28.1zm16-64a5 5 0 1 1 0-2H146v44.1a5 5 0 1 1-2 0V18h-26.1zm102.2 270a5 5 0 1 1 0 2H98v14h-2v-16h124.1zM242 149.9V160h16v34h-16v62h48v48h-2v-46h-48v-66h16v-30h-16v-12.1a5 5 0 1 1 2 0zM53.9 18a5 5 0 1 1 0-2H64V2H48V0h18v18H53.9zm112 32a5 5 0 1 1 0-2H192V0h50v2h-48v48h-28.1zm-48-48a5 5 0 0 1-9.8-2h2.07a3 3 0 1 0 5.66 0H178v34h-18V21.9a5 5 0 1 1 2 0V32h14V2h-58.1zm0 96a5 5 0 1 1 0-2H137l32-32h39V21.9a5 5 0 1 1 2 0V66h-40.17l-32 32H117.9zm28.1 90.1a5 5 0 1 1-2 0v-76.51L175.59 80H224V21.9a5 5 0 1 1 2 0V82h-49.59L146 112.41v75.69zm16 32a5 5 0 1 1-2 0v-99.51L184.59 96H300.1a5 5 0 0 1 3.9-3.9v2.07a3 3 0 0 0 0 5.66v2.07a5 5 0 0 1-3.9-3.9H185.41L162 121.41v98.69zm-144-64a5 5 0 1 1-2 0v-3.51l48-48V48h32V0h2v50H66v55.41l-48 48v2.69zM50 53.9v43.51l-48 48V208h26.1a5 5 0 1 1 0 2H0v-65.41l48-48V53.9a5 5 0 1 1 2 0zm-16 16V89.41l-34 34v-2.82l32-32V69.9a5 5 0 1 1 2 0zM12.1 32a5 5 0 1 1 0 2H9.41L0 43.41V40.6L8.59 32h3.51zm265.8 18a5 5 0 1 1 0-2h18.69l7.41-7.41v2.82L297.41 50H277.9zm-16 160a5 5 0 1 1 0-2H288v-71.41l16-16v2.82l-14 14V210h-28.1zm-208 32a5 5 0 1 1 0-2H64v-22.59L40.59 194H21.9a5 5 0 1 1 0-2H41.41L66 216.59V242H53.9zm150.2 14a5 5 0 1 1 0 2H96v-56.6L56.6 162H37.9a5 5 0 1 1 0-2h19.5L98 200.6V256h106.1zm-150.2 2a5 5 0 1 1 0-2H80v-46.59L48.59 178H21.9a5 5 0 1 1 0-2H49.41L82 208.59V258H53.9zM34 39.8v1.61L9.41 66H0v-2h8.59L32 40.59V0h2v39.8zM2 300.1a5 5 0 0 1 3.9 3.9H3.83A3 3 0 0 0 0 302.17V256h18v48h-2v-46H2v42.1zM34 241v63h-2v-62H0v-2h34v1zM17 18H0v-2h16V0h2v18h-1zm273-2h14v2h-16V0h2v16zm-32 273v15h-2v-14h-14v14h-2v-16h18v1zM0 92.1A5.02 5.02 0 0 1 6 97a5 5 0 0 1-6 4.9v-2.07a3 3 0 1 0 0-5.66V92.1zM80 272h2v32h-2v-32zm37.9 32h-2.07a3 3 0 0 0-5.66 0h-2.07a5 5 0 0 1 9.8 0zM5.9 0A5.02 5.02 0 0 1 0 5.9V3.83A3 3 0 0 0 3.83 0H5.9zm294.2 0h2.07A3 3 0 0 0 304 3.83V5.9a5 5 0 0 1-3.9-5.9zm3.9 300.1v2.07a3 3 0 0 0-1.83 1.83h-2.07a5 5 0 0 1 3.9-3.9zM97 100a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-48 32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm32 48a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm32-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0-32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm32 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16-64a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 96a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-144a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-96 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm96 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16-64a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-32 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zM49 36a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-32 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm32 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zM33 68a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-48a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 240a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16-64a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16-32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm80-176a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm32 48a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0-32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm112 176a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm-16 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zM17 180a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0 16a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm0-32a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16 0a3 3 0 1 0 0-6 3 3 0 0 0 0 6zM17 84a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm32 64a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm16-16a3 3 0 1 0 0-6 3 3 0 0 0 0 6z'%3E%3C/path%3E%3C/svg%3E&#34;)">You read the whole thing.
Back <a class=underline href=#very-top>to the top</a>?</div></main><footer><div class="h-spaced mb-5 mt-12 text-xs text-verysubtle">All content created by Lukas Knuth and licensed under the
<a class="underline text-bold" href=https://creativecommons.org/licenses/by-sa/4.0/>Creative Commons Attribution-ShareAlike 4.0 International License</a>.
All source code is licensed <a class=underline href=https://opensource.org/license/mit>MIT</a>.
Zu <a class="underline text-bold" href=/impressum>Impressum und Datenschutz</a>, vom deutschen Gesetz vorgeschrieben.</div></footer></body></html>